docs/tutorial/codecs.txt

```diff
@@ -0,0 +1,433 @@
+
+           return $person;
+       }
+
+       public function encode($value): MongoDB\BSON\Document
+       {
+           if (! $this->canEncode($value)) {
+               throw UnsupportedValueException::invalidEncodableValue($value);
+           }
+
+           return MongoDB\BSON\Document::fromPHP([
+               '_id' => $value->id,
+               'name' => $value->name,
+           ]);
+       }
+   }
+
+To then use this codec with a collection, specify the ``codec`` option when selecting the collection:
+
+.. code-block:: php
+
+   <?php
+
+   $client = new MongoDB\Client();
+   $collection = $client->selectCollection('test', 'person', [
+       'codec' => new PersonCodec(),
+   ]);
+
+   $person = new Person('Jane Doe');
+   $collection->insertOne($person);
+
+   $person = $collection->findOne();
+
+The example above selects a collection and instructs it to use the ``PersonCodec`` for encoding and decoding documents.
+When inserting data, the ``PersonCodec`` is used to encode the document. When retrieving data, the ``PersonCodec`` is
+used to decode BSON data into a ``Person`` instance. Note that while the ``PersonCodec`` could technically decode any
+BSON document that contains a name field, we wouldn't want to use it for any other documents. Document codecs are meant
+to be used in a collection, or when decoding embedded documents.
+
+When working on a collection with a codec, the codec will only accept and return data of that type for certain
+operations. The ``bulkWrite``, ```findOneAndReplace``, ``insertMany``, ``insertOne``, and ``replaceOne`` operations
+will attempt to encode the given data using the provided codec. Trying to insert or replace a document that cannot be
+encoded will result in an exception. The ``aggregate``, ``find``, ``findOne``, ``findOneAndDelete``,
+``findOneAndReplace``, and ``findOneAndUpdate`` operations will attempt to decode returned documents using the provided
+codec. If the codec does not support the data returned, an exception will be thrown. You can disable codec usage for a
+specific operation or use a different codec (e.g. to decode the result of an aggregation pipeline) by specifying the
+nullable ``codec`` option for the operation. This will override the collection-level codec:
+
+.. code-block:: php
+
+   <?php
+
+   // Returns a document using the default type map
+   $collection->findOne(['name' => 'Jane Doe'], [
+       'codec' => null,
+   ]);
+
+   // Disables codec usage as the aggregate result will have a different format
+   $collection->aggregate([['$group' => [
+       '_id' => '$accountType',
+       'count' => ['$sum' => 1],
+   ]]], [
+       'codec' => null,
+   ]);
+
+Generic Codecs
+--------------
+
+The previous example showed how to define a codec for a specific class. However, sometimes you want to define codecs to
+handle a given type in all documents. For such codecs, you can implement the ``MongoDB\Codec\Codec`` interface. The
```

Comment:
Suggestion:

> The previous example showed how to define a codec for a specific class. However, you may want to create a codec that handles a particular data type in any document. This can be achieved by implementing the ``MongoDB\Codec\Codec`` interface.
>
> The following example defines...

This splits up the intro paragraph with a sentence introducing the example code.

My other motivation was the clarify that we're designing a codec for a particular type in _any_ document vs. "all documents", since you might still want to limit such a codec to certain use cases.